packages/xstate-test/test/coverage.test.ts

```diff
@@ -0,0 +1,389 @@
+                NEXT: 'two'
+              }
+            },
+            two: {},
+            threeMissing: {}
+          }
+        }
+      }
+    });
+
+    const testModel = createTestModel(machine);
+
+    const plans = testModel.getShortestPlans();
+
+    for (const plan of plans) {
+      await testModel.testPlan(plan);
+    }
+
+    expect(
+      testModel
+        .getCoverage(coversAllStates())
+        .filter((c) => c.status !== 'covered')
+        .map((c) => c.criterion.description)
+    ).toMatchInlineSnapshot(`
+      Array [
+        "Visits \\"test.secondMissing\\"",
+        "Visits \\"test.third.threeMissing\\"",
+      ]
+    `);
+
+    expect(() => testModel.testCoverage(coversAllStates()))
+      .toThrowErrorMatchingInlineSnapshot(`
+      "Coverage criteria not met:
+      	Visits \\"test.secondMissing\\"
+      	Visits \\"test.third.threeMissing\\""
+    `);
+  });
+
+  // https://github.com/statelyai/xstate/issues/729
+  it('reports full coverage when all states are covered', async () => {
+    const feedbackMachine = createTestMachine({
+      id: 'feedback',
+      initial: 'logon',
+      states: {
+        logon: {
+          initial: 'empty',
+          states: {
+            empty: {
+              on: {
+                ENTER_LOGON: 'filled'
+              }
+            },
+            filled: { type: 'final' }
+          },
+          on: {
+            LOGON_SUBMIT: 'ordermenu'
+          }
+        },
+        ordermenu: {
+          type: 'final'
+        }
+      }
+    });
+
+    const model = createTestModel(feedbackMachine);
+
+    for (const plan of model.getShortestPlans()) {
+      await model.testPlan(plan);
+    }
+
+    const coverage = model.getCoverage(coversAllStates());
+
+    expect(coverage).toHaveLength(5);
+
+    expect(coverage.filter((c) => c.status !== 'covered')).toHaveLength(0);
+
+    expect(() => model.testCoverage(coversAllStates())).not.toThrow();
+  });
+
+  it('skips filtered states (filter option)', async () => {
+    const TestBug = createTestMachine({
+      id: 'testbug',
+      initial: 'idle',
+      context: {
+        retries: 0
+      },
+      states: {
+        idle: {
+          on: {
+            START: 'passthrough'
+          }
+        },
+        passthrough: {
+          always: 'end'
+        },
+        end: {
+          type: 'final'
+        }
+      }
+    });
+
+    const testModel = createTestModel(TestBug);
+
+    const testPlans = testModel.getShortestPlans();
+
+    const promises: any[] = [];
+    testPlans.forEach((plan) => {
+      plan.paths.forEach(() => {
+        promises.push(testModel.testPlan(plan));
+      });
+    });
+
+    await Promise.all(promises);
+
+    expect(() => {
+      testModel.testCoverage(
+        coversAllStates({
+          filter: (stateNode) => {
+            return stateNode.key !== 'passthrough';
+          }
+        })
+      );
+    }).not.toThrow();
+  });
+
+  // https://github.com/statelyai/xstate/issues/981
```

Comment:
the linked issue doesnt seem to directly relate to the failure in this test - the issue mentions some invalid warnings to be logged when testing machines with final states, but this test fails on the coverage computation
```suggestion
```